File: chess2d/chess/engine/PVSaver.cs
/*-
* Copyright (c) 2003, 2004 Thorsten Greiner
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions
* are met:
* 1. Redistributions of source code must retain the above copyright
*    notice, this list of conditions and the following disclaimer.
* 2. Redistributions in binary form must reproduce the above copyright
*    notice, this list of conditions and the following disclaimer in the
*    documentation and/or other materials provided with the distribution.
*
* THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
* ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
* ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
* FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
* DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
* OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
* HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
* LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
* OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
* SUCH DAMAGE.
*
* $Id: PVSaver.java 2 2007-08-09 07:05:44Z tetchu $
*/
using System;
using System.Collections.Generic;
namespace tgreiner.amy.chess.engine
{
	
	/// <summary> Helper class to save principal variations.
	/// 
	/// </summary>
	/// <author>  Thorsten Greiner
	/// </author>
	public class PVSaver
	{
		/// <summary> Get the ponder move.
		/// 
		/// </summary>
		/// <returns> the ponder move
		/// </returns>
		virtual public int PonderMove
		{
			get
			{
				int[] pv = (int[]) pvs[0];
				
				if (pv.Length > 1)
				{
					return pv[1];
				}
				else
				{
					return 0;
				}
			}
			
		}
		
		/// <summary>List of PVs. </summary>
        private IList<int[]> pvs = new List<int[]>();
		
		/// <summary> Create a PVSaver.</summary>
		public PVSaver()
		{
		}
		
		/// <summary> Ensure that at least <code>size</code> entries in the PV table are
		/// available.
		/// 
		/// </summary>
		/// <param name="size">the size
		/// </param>
		private void  ensureSize(int size)
		{
			while (pvs.Count < size)
			{
				pvs.Add(new int[0]);
			}
		}
		
		/// <summary> 
		/// Record a terminal position.
		/// 
		/// </summary>
		/// <param name="ply">the ply.
		/// </param>
		public virtual void  terminal(int ply)
		{
			ensureSize(ply + 1);
			
			int[] pv = (int[]) pvs[ply];
			
			if (pv.Length != ply)
			{
				pv = new int[ply];
				pvs[ply] = pv;
			}
		}
		
		/// <summary> Record a move.
		/// 
		/// </summary>
		/// <param name="ply">the ply.
		/// </param>
		/// <param name="move">the move.
		/// </param>
		public virtual void  move(int ply, int move)
		{
			ensureSize(ply + 2);
			
			int[] previousPV = (int[]) pvs[ply + 1];
			int[] pv = (int[]) pvs[ply];
			
			int len = System.Math.Max(previousPV.Length, ply + 1);
			
			if (pv.Length != len)
			{
				pv = new int[len];
				pvs[ply] = pv;
			}
			
			pv[ply] = move;
			
			if (previousPV.Length > (ply + 1))
			{
				Array.Copy(previousPV, ply + 1, pv, ply + 1, previousPV.Length - ply - 1);
			}
		}
		
		/// <summary> Get the principal variation.
		/// 
		/// </summary>
		/// <returns> the principal variation
		/// </returns>
		internal virtual int[] getPV()
		{
			return (int[]) pvs[0];
		}
		
		/// <summary> Get the saved PV as String.
		/// 
		/// </summary>
		/// <param name="board">the board.
		/// </param>
		/// <returns> the saved principal variation
		/// </returns>
		public virtual System.String getPV(ChessBoard board)
		{
			System.Text.StringBuilder result = new System.Text.StringBuilder();
			int i;
			int[] pv = (int[]) pvs[0];
			for (i = 0; i < pv.Length; i++)
			{
				int move = pv[i];
				if (!board.isLegalMove(move))
				{
					break;
				}
				result.Append(Move.toSAN(board, move));
				result.Append(" ");
				board.doMove(move);
			}
			for (int j = i - 1; j >= 0; j--)
			{
				board.undoMove();
			}
			return result.ToString();
		}
	}
}

File: chess4d_cpp/chess/engine/PVSaver.cpp
#include "chess/engine/PVSaver.h"

#include <algorithm>

#include "chess/engine/ChessBoard.h"
#include "chess/engine/Move.h"

namespace tgreiner::amy::chess::engine {

int PVSaver::getPonderMove() const
{
    if (pvs_.empty() || pvs_[0].size() <= 1) {
        return 0;
    }

    return pvs_[0][1];
}

void PVSaver::ensureSize(int size)
{
    while (static_cast<int>(pvs_.size()) < size) {
        pvs_.emplace_back();
    }
}

void PVSaver::terminal(int ply)
{
    ensureSize(ply + 1);

    std::vector<int>& pv = pvs_[ply];
    if (static_cast<int>(pv.size()) != ply) {
        pv.assign(static_cast<std::size_t>(ply), 0);
    }
}

void PVSaver::move(int ply, int moveValue)
{
    ensureSize(ply + 2);

    const std::vector<int>& previousPV = pvs_[ply + 1];
    std::vector<int>& pv = pvs_[ply];

    const int len = std::max(static_cast<int>(previousPV.size()), ply + 1);
    if (static_cast<int>(pv.size()) != len) {
        pv.assign(static_cast<std::size_t>(len), 0);
    }

    pv[ply] = moveValue;

    if (static_cast<int>(previousPV.size()) > (ply + 1)) {
        std::copy(previousPV.begin() + ply + 1, previousPV.end(), pv.begin() + ply + 1);
    }
}

std::vector<int> PVSaver::getPV() const
{
    if (pvs_.empty()) {
        return {};
    }

    return pvs_[0];
}

std::string PVSaver::getPV(ChessBoard& board)
{
    static const std::vector<int> emptyPv;

    std::string result;
    int i = 0;
    const std::vector<int>& pv = pvs_.empty() ? emptyPv : pvs_[0];

    for (; i < static_cast<int>(pv.size()); ++i) {
        const int moveValue = pv[i];
        if (!board.isLegalMove(moveValue)) {
            break;
        }

        result += Move::toSAN(board, moveValue);
        result += ' ';
        board.doMove(moveValue);
    }

    for (int j = i - 1; j >= 0; --j) {
        board.undoMove();
    }

    return result;
}

} // namespace tgreiner::amy::chess::engine
